### PR TITLE
idmapping: disallow hostids intersecting subids

### LINKED ISSUE #3416: Bad behavior when UID from the host is in the range mapped into the container
# Required information

 * Distribution: ubuntu 
 * Distribution version: 16.04 LTS
 * The output of "lxc info" or if that fails:
config: {}
api_extensions:
- storage_zfs_remove_snapshots
- container_host_shutdown_timeout
- container_syscall_filtering
- auth_pki
- container_last_used_at
- etag
- patch
- usb_devices
- https_allowed_credentials
- image_compression_algorithm
- directory_manipulation
- container_cpu_time
- storage_zfs_use_refquota
- storage_lvm_mount_options
- network
- profile_usedby
- container_push
- container_exec_recording
- certificate_update
- container_exec_signal_handling
- gpu_devices
- container_image_properties
- migration_progress
- id_map
- network_firewall_filtering
- network_routes
- storage
- file_delete
- file_append
- network_dhcp_expiry
- storage_lvm_vg_rename
- storage_lvm_thinpool_rename
- network_vlan
- image_create_aliases
- container_stateless_copy
- container_only_migration
- storage_zfs_clone_copy
- unix_device_rename
- storage_lvm_use_thinpool
- storage_rsync_bwlimit
- network_vxlan_interface
- storage_btrfs_mount_options
- entity_description
- image_force_refresh
- storage_lvm_lv_resizing
api_status: stable
api_version: "1.0"
auth: trusted
public: false
environment:
  addresses: []
  architectures:
  - x86_64
  - i686
  certificate: |
    -----BEGIN CERTIFICATE-----
    MIIFcTCCA1mgAwIBAgIRAM9mMth1VvI3AbEmfe+qsqkwDQYJKoZIhvcNAQELBQAw
    RDEcMBoGA1UEChMTbGludXhjb250YWluZXJzLm9yZzEkMCIGA1UEAwwbcm9vdEBn
    cmVudGguaW5mby51bmljYWVuLmZyMB4XDTE3MDYxMjEzMTkyMFoXDTI3MDYxMDEz
    MTkyMFowRDEcMBoGA1UEChMTbGludXhjb250YWluZXJzLm9yZzEkMCIGA1UEAwwb
    cm9vdEBncmVudGguaW5mby51bmljYWVuLmZyMIICIjANBgkqhkiG9w0BAQEFAAOC
    Ag8AMIICCgKCAgEA1sUjCDiVObv1xbzx34I4zw3aZjZnM1PVmjb0EuX5KjLEh1WX
    057BPhIPk4dJDtlB8uiFfUXrE6VQAncAWQTDpgzO0Gg7ZeXJF24bEoywxk4TQYGV
    D6CHGpDwFM1gm4pXwNtygc0TxnFdGbg8y7MUw2ANqWJyFuro3Z04lc4WWFk8jmuR
    D9mRX93I2zlpfq3keM7oclohrZbb6/cNyAZ0eLTgBqYi/BUH0r1wTv7c+2JFYWRa
    9MSTf/u6d4pzxZbm4/9r4jRGYi1YVK4nBQsAG8S6aoTeueFT72ec5UYgPPt7l2LG
    WaICfPefgL+230AklvfsyuNU4Ljd2uKlpMk+C/RKLxvabCOwnqW1/SpQKsWR1UeI
    VTvOmBGE4bnySrVI7joakSJM8nc/oZDWdHm1kX7ojc3e20Vb4Uryf65qdRyuEOy9
    4PlcMkR04bc3SDX3RB+ynP/rG1ZucdoiMnb5AzNICpIRhl3k8jE0onMCXcpwA/QX
    cdiVgu0D3pmgzZuJ0Sid1UcB+0iS9h2tisaPjHCqc+huKatGBXxlfIWBFzlkBWlf
    2H5m8BBfguMH5garQDsWlKdlwuiwuAOy8ndHHFQ3NaRNuJWoCuOsy8+xlnVqnvDD
    NbFd1nMIDZS8AWy1xXnA9Vc3FqrFoEwwj2i+fPqq+/Fn2Aqg4jmzhTrF9mkCAwEA
    AaNeMFwwDgYDVR0PAQH/BAQDAgWgMBMGA1UdJQQMMAoGCCsGAQUFBwMBMAwGA1Ud
    EwEB/wQCMAAwJwYDVR0RBCAwHoIWZ3JlbnRoLmluZm8udW5pY2Flbi5mcocECtcL
    7zANBgkqhkiG9w0BAQsFAAOCAgEAkUJqfgMWfmPhupaNh4aE9MGW5ObXc23+6aVR
    61BRZ0ABXSEsoPK27Nk/Dz61lc2LoFPfy2oKJfhdHkSIo/uQejKSNt18LB+soSfl
    pr7hiyr+2+nTUwZwsIST7yYUzMXmQeCRneSLOrMSFdeQu0B6k5nAODZqYpT6fDB0
    2Tyvowl1QBVVyvZ+s79/aqt6d1oMBdLB0vv2kzs7VBbkSGQRugu9v88xKRDYnVkI
    c3VlszCfTOo08JpBWyKTZovvqN/YPQLtyVj7xkU4PYlgv2jeO9AkipRGdZJG9E6i
    qVw1K4I3108kAODKbCOixmd6SJ4Mjn6LB5aUMnmpwaiT1huhZRSJmfOprvp0q5SF
    JNCuv7qv6Nx4iog+89mazkUW/ug7pW65hSM6vG6EL1Co8M7EnFBXqYTfh6RWL1DY
    J3C5aWulPgqDuGd+a3v5u1gWEd0zeUTaVxGGV0e1nPvArXX1AHT9GC0Bb8rLHSNE
    RRGtTHtxHVmLHj+5pNsRHxsYI4yXF7TweWOKzmDoFORhw8QY+OZp4f5ZGiTsPh3r
    dyJc2HIqGRg/CBC1mBAdyI7UqUAc91GEXgVlNZg//em1NFo//FZTZuM6CI0O8DBN
    d2B/m1+SasJhccjN72cISdstd8XCRYHGNYbZmSoWUvzjEn0sKyLCdxGemLg1TVqj
    2NSAGWQ=
    -----END CERTIFICATE-----
  certificate_fingerprint: f2befaf411895f024ef6bf31c159f99a4b4c16e4cd656419924d9c06f0eae775
  driver: lxc
  driver_version: 2.0.7
  kernel: Linux
  kernel_architecture: x86_64
  kernel_version: 4.4.0-79-generic
  server: lxd
  server_pid: 8188
  server_version: "2.14"
  storage: zfs
  storage_version: 0.6.5.6-0ubuntu16

# Issue description

I wanted to use GUI apps in my containers, I followed the great howto
here
https://blog.simos.info/how-to-run-graphics-accelerated-gui-apps-in-lxd-containers-on-your-ubuntu-desktop/

However, on my system, it happens that my uid is greater than 65535:

    $ id -u $USER
    119373

When using

    lxc config set guiapps raw.idmap "uid $UID 1000"

in the container, the mapped id is 19737(not 119737), like for .Xauthority :

    $ ls -la /home/ubuntu/.Xauthority
    -rw------- 1 19373 ubuntu 67 Jun 14 07:46 /home/ubuntu/.Xauthority

# Steps to reproduce

 1. create an user with UID larger than 65535 (more than 16 bits)
 2. add this user in lxd group
 3. try to map UID in one container

### PR DESCRIPTION
**Pull request by brauner**
Our code has explicitly disallowed this for a long time but somehow we have
never correctly errored out in these cases like we should.
After having discussed this with Eric at Plumbers I think this should be
considered a security liability. The problem I see with this is that it
accidently allows users to escalate a container that is unprivileged to a
privileged container without their knowledge. The most obvious case is when
the allocated range of subids is accidently 0-65536 which is rare but basically
runs afoul of creating an unprivileged container but there are more subtle
problems when using "{uid,gid,both} \<id\> \<id\>" directly so I prefer to disallow
this for now. Especially since this feature is currently not very useful since
the maximum number of allowed idmappings is currently capped at five so this
would overflow quite easy.

Closes #3416.

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

### CODE REVIEW

File: test/suites/idmap.sh [hunk 1]
```diff
@@ -219,4 +221,37 @@ EOF
...
+    UID_BASE=0
+    GID_BASE=0
+
+    mv /etc/subuid /etc/subuid_backup
+    mv /etc/subgid /etc/subgid_backup
+
+    echo "lxd:"$UID_BASE":$UIDs" > /etc/subuid
+    echo "lxd:"$UID_BASE":$GIDs" > /etc/subgid
+    echo "root:"$GID_BASE":$UIDs" > /etc/subuid
+    echo "root:"$GID_BASE":$GIDs" > /etc/subgid
```
Comment (stgraber) [inline test/suites/idmap.sh]: We absolutely can't do that kind of stuff in the testsuite, a failure in the middle of this code would break the system and when running multiple testsuites in parallel, you'll end up with undefined results.

We can't have the testsuite modify system files like that.

**Comment by stgraber** [comment]
I'll have to think through this some more tomorrow. I do seem to remember there being very much legitimate use cases for the original bug report and so just denying it doesn't feel lite the right option, but I need to dig into this some more myself.

File: test/suites/idmap.sh (hunk 1 above)
Comment (brauner) [reply test/suites/idmap.sh]: ok

**Comment by brauner** [comment]
Yeah, I'm not denying that there are legitimate use cases I'm just concerned about edge cases and furthermore I'm unsure whether the benefits of allowing this licenses the increased complexity of the algorithm to handle these cases at this point in time when we are limited wrt to the useable mappings. The last two days I tried to handle cases like:

```
uid 1000 1000
uid  883 1001
```

which e.g. would (given a range of e.g. `999-1000000000`) translate to roughly something like

```
lxc.idmap = u     0  999   1
lxc.idmap = u  1000 1000   1
lxc.idmap = u   883 1001   1
lxc.idmap = u     1 1002 883
lxc.idmap = u   884 1885 116
lxc.idmap = u  1001 2001 end

```

```
uid 1000 1000
uid 1001 1001
```
which makes a whole-punching algorithm a real pain. :)

File: lxd/container_lxc.go [hunk 2]
```diff
@@ -642,12 +642,22 @@ func findIdmap(state *state.State, cName string, isolatedStr string, configBase
 		return nil, 0, err
 	}
 
+	// Check that this is not an identical mapping.
+	for _, v := range state.OS.IdmapSet.Idmap {
+		if v.Nsid == v.Hostid {
+			return nil, 0, idmap.ErrHostIdIsSubId
+		}
+	}
```
Comment (stgraber) [inline lxd/container_lxc.go]: Per our call, we should drop that.

**Comment by stgraber** [comment]
jenkins: test this please

**Comment by jfgibbins** [comment]
The disallow would only apply when the ID your mapping from the host, is one that would already be in the container mapping, ie container is 100000-165535, and someone wants to remap 101000 at a new location?  As long as the mapped ID isn't already in the container to begin with, mapping will be allowed without incident?

**Comment by brauner** [comment]
@jfgibbins, correct. :)

**Comment by brauner** [comment]
We'll certainly will implement a proper algorithm handling that case as well but that will take more thought.
